### PR TITLE
[docs] Migrate Plugin contribution docs

### PR DESCRIPTION
**Pull request by vmdef**
The remaining plugin contribution documentation is being migrated.

<a href="https://gitpod.io/#https://github.com/moodle/devdocs/pull/696"></a>

### CODE REVIEW

**Comment by vmdef** [comment]
Hello @sarjona, would you be able to review the final migration of the plugin contribution documentation? Thank you.

File: general/community/plugincontribution/adoption.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+title: Plugins adoption programme
+sidebar_position: 9
+tags:
+  - Guidelines for contributors
+  - Plugins
+  - Plugin documentation
+---
+The Plugins adoption programme is a process for making it clear that a plugin is orphaned and is looking for a new maintainer. The programme is one of the mechanisms helping to minimise the risks of relying on additional plugins. The programme helps to find new maintainers for plugins whose original authors can't work on the plugin fully any more.
+
+The programmes started in 2014 by announcing it in [a forum post](https://moodle.org/mod/forum/discuss.php?d=260354#p1128482) and since then, many plugins found their new maintainers since then through it.
```
Comment (sarjona) [inline general/community/plugincontribution/adoption.md:11]:
```suggestion
:::info

The programmes started in 2014 by announcing it in [a forum post](https://moodle.org/mod/forum/discuss.php?d=260354#p1128482) and since then, many plugins found their new maintainers since then through it.

:::
```

File: general/community/plugincontribution/adoption.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+### Motivation
+
+Having an additional plugin installed at a Moodle site is always a risk. One of the essential aspects (apart from the code quality itself) that potential plugin users have to consider is how well the plugin is being maintained. Does the maintainer release regular updates and bug fixes? Is the plugin updated every six months for the new Moodle major release? Is there a place to report bugs and feature requests? And when reported, are they reflected?
+
+It's not that difficult to write a new Moodle plugin these days. Many students do that as a part of their school or thesis projects, for example. But can one rely on the author of plugin to provide sufficient (or at least some) support for it? To be a responsible maintainer of a plugin is much harder than to be an author of it. Many maintainers work on their plugins in their free time. And even if they are lucky enough to be paid for doing that, it's just time consuming (as everything).
+
+At certain moment, maintainers can realise they are not able to give enough love to their plugins any more. In the essay The Cathedral and the Bazaar, Eric Steven Raymond says
```
Comment (sarjona) [inline general/community/plugincontribution/adoption.md:19]:
```suggestion
At certain moment, maintainers can realise they are not able to give enough love to their plugins any more. In the essay "The Cathedral and the Bazaar", Eric Steven Raymond says:
```

File: general/community/plugincontribution/adoption.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+:::note
+
+When you lose interest in a program, your last duty to it is to hand it off to a competent successor.
```
Comment (sarjona) [inline general/community/plugincontribution/adoption.md:23]:
```suggestion
> When you lose interest in a program, your last duty to it is to hand it off to a competent successor.
```

File: general/community/plugincontribution/adoption.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+### Motivation
+
+Having an additional plugin installed at a Moodle site is always a risk. One of the essential aspects (apart from the code quality itself) that potential plugin users have to consider is how well the plugin is being maintained. Does the maintainer release regular updates and bug fixes? Is the plugin updated every six months for the new Moodle major release? Is there a place to report bugs and feature requests? And when reported, are they reflected?
+
+It's not that difficult to write a new Moodle plugin these days. Many students do that as a part of their school or thesis projects, for example. But can one rely on the author of plugin to provide sufficient (or at least some) support for it? To be a responsible maintainer of a plugin is much harder than to be an author of it. Many maintainers work on their plugins in their free time. And even if they are lucky enough to be paid for doing that, it's just time consuming (as everything).
+
+At certain moment, maintainers can realise they are not able to give enough love to their plugins any more. In the essay The Cathedral and the Bazaar, Eric Steven Raymond says
+
+:::note
```
Comment (sarjona) [inline general/community/plugincontribution/adoption.md:21]:
```suggestion
```

File: general/community/plugincontribution/adoption.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+When you lose interest in a program, your last duty to it is to hand it off to a competent successor.
+
+:::
```
Comment (sarjona) [inline general/community/plugincontribution/adoption.md:25]:
```suggestion
```

File: general/community/plugincontribution/adoption.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+It's not that difficult to write a new Moodle plugin these days. Many students do that as a part of their school or thesis projects, for example. But can one rely on the author of plugin to provide sufficient (or at least some) support for it? To be a responsible maintainer of a plugin is much harder than to be an author of it. Many maintainers work on their plugins in their free time. And even if they are lucky enough to be paid for doing that, it's just time consuming (as everything).
+
+At certain moment, maintainers can realise they are not able to give enough love to their plugins any more. In the essay The Cathedral and the Bazaar, Eric Steven Raymond says
+
+:::note
+
+When you lose interest in a program, your last duty to it is to hand it off to a competent successor.
+
```
Comment (sarjona) [inline general/community/plugincontribution/adoption.md:24]:
```suggestion
```

File: general/community/plugincontribution/pluginsdirectoryapi.md
Comment (sarjona) [inline general/community/plugincontribution/pluginsdirectoryapi.md:1]: As suggested with the qaprechecks, shouldn't better to create a folder general/community/plugincontribution/directory and move this file there (general/community/plugincontribution/directory/api.md) ? That way, it will be clearer that this section is related to the Plugins directory.

If you create this folder, you'll also need to move general/community/plugincontribution/pluginsdirectory.md to general/community/plugincontribution/directory/index.md

File: general/community/plugincontribution/pluginsdirectoryapi.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+                'smurfresult' => new external_value(PARAM_TEXT, 'Code prechecks results summary'),
+            ])
+        ),
+    ])
+);
+```
+
+#### Example cURL client fetching the list of maintained plugins
+
+```bash
```
Comment (sarjona) [inline general/community/plugincontribution/pluginsdirectoryapi.md:95]:
```suggestion
```bash title="Example cURL client fetching the list of maintained plugins"
```

File: general/community/plugincontribution/pluginsdirectoryapi.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                'downloadurl' => new external_value(PARAM_URL, 'Download URL'),
+                'viewurl' => new external_value(PARAM_URL, 'View URL'),
+                'smurfresult' => new external_value(PARAM_TEXT, 'Code prechecks results summary'),
+            ])
+        ),
+    ])
+);
+```
+
+#### Example cURL client fetching the list of maintained plugins
```
Comment (sarjona) [inline general/community/plugincontribution/pluginsdirectoryapi.md:93]:
```suggestion
```

File: general/community/plugincontribution/pluginsdirectoryapi.md [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
```
Comment (sarjona) [inline general/community/plugincontribution/pluginsdirectoryapi.md:94]:
```suggestion
```

File: general/community/plugincontribution/pluginsdirectoryapi.md [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    'viewurl' => new external_value(PARAM_URL, 'View URL'),
+    'warnings' => new external_multiple_structure(
+        new external_value(PARAM_RAW, 'Validation warnings')
+    ),
+]);
+```
+
+#### Example CLI script to release a new version of a plugin
+
+```bash
```
Comment (sarjona) [inline general/community/plugincontribution/pluginsdirectoryapi.md:164]:
```suggestion
```bash title="Example CLI script to release a new version of a plugin"
```

File: general/community/plugincontribution/pluginsdirectoryapi.md [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    'timecreated' => new external_value(PARAM_INT, 'Timestamp of version release'),
+    'downloadurl' => new external_value(PARAM_URL, 'Download URL'),
+    'viewurl' => new external_value(PARAM_URL, 'View URL'),
+    'warnings' => new external_multiple_structure(
+        new external_value(PARAM_RAW, 'Validation warnings')
+    ),
+]);
+```
+
+#### Example CLI script to release a new version of a plugin
```
Comment (sarjona) [inline general/community/plugincontribution/pluginsdirectoryapi.md:162]:
```suggestion
```

File: general/community/plugincontribution/pluginsdirectoryapi.md [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
```
Comment (sarjona) [inline general/community/plugincontribution/pluginsdirectoryapi.md:163]:
```suggestion
```

File: general/community/plugincontribution/pluginsdirectoryapi.md [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+ENDPOINT_REST="${HOST}/webservice/rest/server.php"
+ENDPOINT_UPLOAD="${HOST}/webservice/upload.php"
+
+ITEMID=$(${CURL} -F data=@${ZIP} "${ENDPOINT_UPLOAD}?token=${TOKEN}" | jq --raw-output '.[0].itemid')
+
+${CURL} ${ENDPOINT_REST} --data-urlencode "wstoken=${TOKEN}" --data-urlencode "wsfunction=${FUNCTION}" --data-urlencode "moodlewsrestformat=json" \
+    --data-urlencode "frankenstyle=${PLUGIN}" --data-urlencode "zipdrafitemtid=${ITEMID}" | jq
+```
+
+#### GitHub Actions
```
Comment (sarjona) [inline general/community/plugincontribution/pluginsdirectoryapi.md:186]:
```suggestion
### GitHub Actions
```

File: .vscode/settings.json [hunk 14]
```diff
@@ -1,4 +1,7 @@
 {
     "stylelint.packageManager": "yarn",
-    "cSpell.enabled": true
+    "cSpell.enabled": true,
+    "cSpell.words": [
+        "prechecks"
```
Comment (sarjona) [inline .vscode/settings.json:5]: Why is this change required?

Comment (vmdef) [reply .vscode/settings.json:5]: I received multiple cSpell errors regarding unrecognized words while attempting to commit. I assumed that it was necessary to include them in this file.

Comment (vmdef) [reply .vscode/settings.json:5]: I have just realized that this is a file for the editor. It seems that I may have created it unconsciously. I'll revert this change.

File: general/community/plugincontribution/checklist.md [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+- All styles.css files from all plugins are concatenated, cached and served as a one big resource on all pages of the given Moodle installation.
+- For that reason, plugins must use properly namespaced selectors so that their style sheets can be safely combined with others without affecting other pages and elements beyond the plugin's scope.
+- Plugin specific CSS selectors are needed to make sure that your styling does not accidentally affect other parts of Moodle outside your plugin scope.
+
+:::tip Example
+
+Instead of the selector `.contentarea` it is better to use something like `.path-mod-mymodule .contentarea` as the `.path-\*` classes are automatically added by the Moodle core renderers to the HTML `<body>` tag.
+
+ :::
```
Comment (sarjona) [inline general/community/plugincontribution/checklist.md:146]:
```suggestion
:::
```

File: general/community/plugincontribution/codeprechecks.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+Plugin version with no detected errors or warnings has a label like this displayed:
+
+![Plugin code prechecks success](_codeprechecks/plugin-codeprechecks-success.png)
+
+If there are some formal errors or warnings detected, a label like this is displayed:
+
+![Plugin code prechecks error](_codeprechecks/plugin-codeprechecks-error.png)
+
+::: info Labels information
```
Comment (sarjona) [inline general/community/plugincontribution/codeprechecks.md:22]:
```suggestion
:::info Labels information
```

**Review by sarjona — CHANGES_REQUESTED** [review]
Hola @vmdef!
Thanks for reviewing the doc and including the suggestions! As you'll see, there are only a couple of minor things to review (and then it will be ready to be merged!).

**Comment by vmdef** [comment]
Hi @sarjona!
Thank you for your prompt review and patience. I have incorporated the suggested modifications into the existing commit "Apply suggestions from code review" (https://github.com/moodle/devdocs/pull/696/commits/85683cc2937de4322549e835943d5b67226a7491) in order to maintain a cleaner history.

**Review by sarjona — APPROVED** [review]
Hola Víctor!
Thanks!! These pages look great and it's great they have been migrated. Well done!  <3
